### PR TITLE
A32: mention circuit breaking atomics are shared globally for entire client process

### PR DESCRIPTION
**Pull request by voidzcy**

### CODE REVIEW

File: A32-xds-circuit-breaking.md [hunk 1]
```diff
@@ -53,13 +53,23 @@ circuit breakers. Users can effectively turn circuit breaking off by setting
 limits to very high values, for example, to 
 `std::numeric_limits<uint32_t>::max()`.
 
+## Related Proposals: 
+
+This proposal builds on the earlier xDS work described in
+- [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+
+Related design changes made after this proposal in
+- [A37: xDS Aggregate and Logical DNS Clusters](https://github.com/grpc/proposal/pull/216)
```
Comment (markdroth) [inline A32-xds-circuit-breaking.md:62]: I think the change here can be explained without actually referencing A37 at all.  I think the only thing we really need to say is that we want to be consistent with Envoy, where the circuit breaking counts are tracked on a per-process basis, not a per-client basis.  So we want the counts to be tracked the same way in gRPC.

File: A32-xds-circuit-breaking.md [hunk 2]
```diff
@@ -93,16 +103,46 @@ The EDS LB policy will use this limit as the upper bound for the maximum number
...
 cluster-level load reports and reported to the load reporting server.
 
+Recent changes to the load balancing policy hierarchy in gRPC split up the EDS
+LB policy into `xds_cluster_resolver_experimental` policy and
+`xds_cluster_impl_experimental` policy. The latter is the place where circuit
+breaking is implemented. See more details in [A37: xDS Aggregate and Logical DNS Clusters](https://github.com/grpc/proposal/pull/216).
+
+### Per cluster requests aggregation
+
+Although with [gRPC A31: xDS RouteActions Support](https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md)
```
Comment (markdroth) [inline A32-xds-circuit-breaking.md:123]: I think this paragraph can be removed.

File: A32-xds-circuit-breaking.md [hunk 3]
```diff
@@ -93,16 +103,46 @@ The EDS LB policy will use this limit as the upper bound for the maximum number
...
+the associated cluster. The counter will be incremented when a request is
+started on a subchannel and decremented when the request is complete. 
 When the picker is going to start a call, it will check whether the current
 number of in-flight requests is lower than the configured limit. If so, the 
 picker will do the pick as usual, returning a subchannel. If not, it will 
 fail the call with status UNAVAILABLE. Such failed calls will not be retried,
 but they will be recorded to the `total_dropped_requests` counts in 
 cluster-level load reports and reported to the load reporting server.
 
+Recent changes to the load balancing policy hierarchy in gRPC split up the EDS
```
Comment (markdroth) [inline A32-xds-circuit-breaking.md:116]: This paragraph isn't needed.  We don't need to update this gRFC to reflect what comes after it, just to reflect the state when we implemented it.  Subsequent designs will always change it, and we don't go back and retrofit all previous gRFCs when that happens, so I'd rather not start giving people false hope that we'll do that. :)

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

File: A32-xds-circuit-breaking.md (hunk 2 above)
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Removed.

File: A32-xds-circuit-breaking.md (hunk 1 above)
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Deleted. So I am only adding a single paragraph to mention circuit breaking is a stateful feature, so the state should be shared for channels in the entire process. See the added section below.

File: A32-xds-circuit-breaking.md (hunk 3 above)
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Deleted.

File: A32-xds-circuit-breaking.md [hunk 4]
```diff
@@ -53,13 +53,23 @@ circuit breakers. Users can effectively turn circuit breaking off by setting
...
 
+### Threshold enforcement in LB policy
+
 With the introduction of [gRFC A28: xDS Traffic Splitting and Routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md), 
-it became possible for there to be more than one CDS LB policy instance for a 
-given cluster (e.g., if one route pointed to a given cluster and a different 
-route split traffic between that cluster and other clusters).  However, after 
-implementing the changes described in [gRPC A31: xDS RouteActions Support](https://github.com/grpc/proposal/pull/192),
+it became possible for each client channel to have more than one CDS LB policy 
+instances for a given cluster (e.g., if one route pointed to a given cluster
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Well, it is saying "more than one" :)

File: A32-xds-circuit-breaking.md [hunk 5]
```diff
@@ -55,11 +55,13 @@ limits to very high values, for example, to
...
```
Comment (markdroth) [inline A32-xds-circuit-breaking.md:62]: s/instances/instance/

File: A32-xds-circuit-breaking.md [hunk 6]
```diff
@@ -93,16 +95,28 @@ The EDS LB policy will use this limit as the upper bound for the maximum number
...
 cluster-level load reports and reported to the load reporting server.
 
+### Globally shared atomics for outstanding requests
+
+In Envoy circuit breakers are applied to the entire traffic of 
+the process sent to upstream clusters as all traffic goes through the sidecar
+proxy. To achieve the same behavior, counters used to aggregate outstanding
+requests should be globally shared for channels in the client process.
+Therefore, we will use a global map to hold call counters per {cluster, EDS
+service name} pair. Counters reference counted and lazily created upon being
```
Comment (markdroth) [inline A32-xds-circuit-breaking.md:115]: s/Counters/Counters are/

**Review by markdroth — APPROVED** [review]
This looks great, modulo a couple of minor grammar nits.  Thanks for doing this!

File: A32-xds-circuit-breaking.md (hunk 5 above)
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Well, shouldn't the noun following "more than one" be plural? I don't find an authoritative answer. Anyway, I am trusting you :)

File: A32-xds-circuit-breaking.md (hunk 6 above)
Comment (voidzcy) [reply A32-xds-circuit-breaking.md]: Fixed, thanks.

**Pull request merged**